Schema enforcement use-case docs (#63)
kroxylicious/kroxylicious.github.io: _use_cases/schema_validation_and_enforecement.markdown

by the producer and the expectations of the consumer, problems will result.  The consuming application could take the wrong business
action or the consuming application may fail leading to complete failure of the system.  This problem is known as *poison message*
scenario.

If there is only a small number of applications, perhaps all managed by the same team, it is possible for the developers to use
informal agreements about what message format will be used on which topic.  However, as the use of Kafka grows within the organization,
the amount of effort required to keep all the formats in agreement grows.  Mistakes, and system downtime, become inevitably.

| ![image]({{'/assets/pages/images/schema_validation_problem.png' | absolute_url}}){:width="100%"} |
|:-----------------------------------------------------------------:|
|    *Problem: Poison message leading to consumer crash*     |

##### Schemas to the rescue

To overcome the problem, the Kafka **client** ecosystem supports _schemas_.  Schemas provide a programmatic description
of the message format.

> Producers use a schema to ensure that they produce messages that adhere to the agreed message format.  The consumer

Reviewer: My mental model was always
> producers use a schema to advertise the format of the messages they (are supposed to) produce

Which is where a proxy would step in to ensure the compliance with the schema.

responsibility on the producing applications to configure and use schemas correctly.  There's nothing in the system to
stop, say, a misconfigured producing application using the wrong schema or failing to use schemas at all.

In Kafka, **brokers** have no knowledge of format of the messages (to the broker, a message is just opaque bytes) and they
have no knowledge of the schema.  The brokers cannot help us enforce that producing applications use schemas and use them
correctly.

#### Kroxylicious Record Validation

The Kroxylicious Record Validation Filter provides a solution to the problem. 

The filter intercepts the produce requests sent from producing applications and subjects them to validation. If
validation fail, the product request is rejected.  The producing application receives an error response.  The broker
does not receive the rejected records.  In this way, one can organize that a poison message never enters the system,
even if producing applications are misconfigured.

The filter currently supports two types of validation:

* _Schema validation_[^1] validates the content of the record against a schema.  

> * _SyntacticallyCorrectJson_ validation ensures the producer is producing messages that contain syntactically valid

Author: Maybe we shouldn't mention `SyntacticallyCorrectJson` in the use-case documentation at all.   I think it was really a POC that helped us prove the idea.

Reviewer: Its one of those things where it helps the unstructured JSON crowd. So if we are taking a binary view and pushing schemas everywhere then I don't think there is any value in mentioning it here.  I'd be quite happy with something like 
> The filter currently supports two modes of operation
> 1. _Schema validation_[^1] validates the content of the record against a schema.   - For topics which have an entry in the centralised schema registry
> 2. _SyntacticallyCorrectJson_ validation ensures the producer is producing messages that contain syntactically valid JSON - For topics which do not have registered schemas.  

Having mentioned the 2nd option I don't think we need to go much further.

  JSON.

Validations may be configured on a per-topic basis.

| ![image]({{'/assets/pages/images/schema_validation_solution.png' | absolute_url}}){:width="100%"} |
|:-----------------------------------------------------------------:|
|    *Proxy configured for schema enforcement*     |

> 

Author: I'd like to add a sequence diagram here to explain the error flow.

---
name: Schema Validation and Enforcement
---

### Schema Validation and Enforcement

#### Why

> In Kafka, producing applications transfer messages to consuming applications.  In order for messages to be transferred successfully

Reviewer 2: ```suggestion
In Apache Kafka, producing applications transfer messages to consuming applications.  In order for messages to be transferred successfully
```

> producer and consumer need to agree on the format of the message being transferred. If there is a mismatch between the format sent

Reviewer 2: ```suggestion
producers and consumers need to agree on the format of the message being transferred. If there is a mismatch between the format sent
```

by the producer and the expectations of the consumer, problems will result.  The consuming application could take the wrong business

> action or the consuming application may fail leading to complete failure of the system.  This problem is known as a *poison message*

Reviewer 2: ```suggestion
action or the consuming application may fail to process the message at all, leading to complete failure of the system.  This problem is known as a *poison message*
```

decode the incoming message with the certainty that it conforms to the expected format.

The schemas themselves could be packaged as part of the application.  Whilst the approach is functional, it is quite
inflexible as it means all applications need to be redeployed every time a schema changes.  This becomes a challenge
as the number of applications in the system grows.  To counter this challenge organizations often choose to deploy a
*Schema Registry*. 

##### Schema Registry

The role of the  **Schema Registry**  is to maintain a library of schemas that are in-use within an organisation.

Producer applications call the registry to retrieve the schema for the messages they need to send.

Consuming applications use information embedded in the incoming message to be able to retrieve the right schema from
the registry.  This allows the consumer to decode the message.

| ![image]({{'/assets/pages/images/schema_validation_schema_registry.png' | absolute_url}}){:width="100%"} |
|:-----------------------------------------------------------------:|
|    *Kafka applications making use of a Schema Registry*     |

> #### So, what's the problem?

Reviewer 3: Maybe we should also point out that maybe not all clients are schema-aware, we may have technology that wants to emit JSON without awareness of how to encode apicurio identifiers into the message. The schema validation check can enforce the schema and prevent invalid messages from these clients entering the topic.

Reviewer 3: might just muddy the usecase though

Author: I think we should go with it as it stands.  I'm sure we'll want to improve the use-case documentation as we grow the feature.

poison message to be introduced into the system is greatly diminished.  However, the issue is that there's significant
responsibility on the producing applications to configure and use schemas correctly.  There's nothing in the system to
stop, say, a misconfigured producing application using the wrong schema or failing to use schemas at all.

In Kafka, **brokers** have no knowledge of format of the messages (to the broker, a message is just opaque bytes) and they
have no knowledge of the schema.  The brokers cannot help us enforce that producing applications use schemas and use them
correctly.

#### Kroxylicious Record Validation

The Kroxylicious Record Validation Filter provides a solution to the problem. 

The filter intercepts the produce requests sent from producing applications and subjects them to validation. If
validation fail, the product request is rejected and the producing application receives an error response.  The broker
does not receive the rejected records.  In this way, one can organize that a poison message never enters the system,
even if producing applications are misconfigured.

The filter currently supports two modes of operation:

> 1. Schema validation[^1] validates the content of the record against a schema. Use this for topics which have an entry in

Reviewer 3: We are also enabling another potential usecase, the user can use a schema from a Registry to validate json blobs being pushed to kafka by a client with no knowledge of Schemas. So we can use technology that is not apicurio integrated, but still prevent invalid JSON from entering the topic.
